# Тестирование АВЛ, рандомизированных и расширяющихся деревьев поиска

In [1]:
import random
import sys

from algorithms_py.trees import TreeAVL, TreeRandom, TreeSplay

sys.setrecursionlimit(20000)

Cлучайным образом выполняет операции добавления, поиска и удаления элемента в соотношении "weights". Поиск и удаление осуществляются на присутствующих в дереве элементах. Таким образом, в процессе еще происходит обход дерева.

In [2]:
def test(tree_type, N, weights=[0.8, 0.1, 0.1], shuffle=False):
    tree = tree_type()
    array = [(i, random.randint(0, 999)) for i in range(N)]
    array_used = []
    if shuffle:
        random.shuffle(array)
    count = 0
    i = 0
    f = 0
    r = 0
    for k in random.choices(['i', 'f', 'r'], weights=weights, k=N*100):
        if array:
            count += 1
            if k == 'i':
                key, value = array.pop(0)
                tree.insert(key, value)
                array_used.append(key)
                i += 1
            elif k == 'f':
                if array_used:
                    key = random.choice(array_used)
                    tree.has_key(key)
                    f += 1
            else:
                try:
                    if array_used:
                        key = random.choice(array_used)
                        tree.remove(key)
                        r += 1
                # если числа нет в дереве
                except KeyError:
                    pass
        if not array:
            break
    #print(count, i, f, r)
    print('Обход дерева:', list(tree)[:10])
    print('Количество узлов:', len(list(tree)))
    return tree

## Тест №1

- Входной массив отсортирован
- Операций добавления больше, чем операций поиска и удаления
- N = 50000

In [3]:
N = 50000
weights = [5, 1, 1]

In [4]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeAVL, N, weights=weights)

Обход дерева: [(5, 706), (15, 53), (23, 103), (27, 35), (29, 568), (30, 897), (35, 99), (42, 334), (50, 695), (53, 597)]
Количество узлов: 41641


<TimeitResult : 3.38 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

In [5]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeRandom, N, weights=weights)

Обход дерева: [(1, 297), (15, 905), (18, 70), (20, 373), (21, 428), (26, 607), (30, 103), (32, 537), (34, 463), (36, 505)]
Количество узлов: 41554


<TimeitResult : 3.84 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

In [6]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeSplay, N, weights=weights)

Обход дерева: [(0, 406), (5, 608), (9, 690), (21, 682), (22, 758), (23, 552), (26, 994), (34, 647), (36, 898), (37, 499)]
Количество узлов: 41642


<TimeitResult : 8.76 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

## Тест №2

- Входной массив отсортирован
- Операций поиска больше, чем операций добавления и удаления
- N = 50000

In [7]:
N = 50000
weights = [2, 5, 1]

In [8]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeAVL, N, weights=weights)

Обход дерева: [(7, 674), (16, 450), (70, 588), (91, 55), (146, 228), (152, 543), (172, 635), (182, 456), (190, 232), (217, 398)]
Количество узлов: 33350


<TimeitResult : 4.24 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

In [9]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeRandom, N, weights=weights)

Обход дерева: [(86, 126), (118, 330), (123, 983), (125, 71), (146, 825), (170, 607), (179, 831), (209, 343), (260, 98), (274, 805)]
Количество узлов: 33481


<TimeitResult : 7.45 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

In [10]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeSplay, N, weights=weights)

Обход дерева: [(67, 800), (84, 581), (89, 469), (96, 884), (151, 850), (193, 933), (220, 859), (231, 106), (262, 198), (272, 418)]
Количество узлов: 33150


<TimeitResult : 6.79 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

## Тест №3

- Входной массив перемешан
- Операций добавления больше, чем операций поиска и удаления
- N = 50000

In [11]:
N = 50000
weights = [5, 1, 1]

In [12]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeAVL, N, weights=weights, shuffle=True)

Обход дерева: [(1, 725), (2, 732), (4, 738), (5, 177), (6, 797), (7, 555), (8, 596), (10, 344), (13, 313), (14, 2)]
Количество узлов: 41772


<TimeitResult : 3.33 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

In [13]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeRandom, N, weights=weights, shuffle=True)

Обход дерева: [(0, 129), (1, 869), (2, 928), (3, 946), (5, 95), (6, 735), (7, 583), (8, 332), (9, 205), (11, 850)]
Количество узлов: 41754


<TimeitResult : 8.85 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

In [14]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeSplay, N, weights=weights, shuffle=True)

Обход дерева: [(1, 185), (3, 339), (4, 101), (5, 660), (6, 47), (8, 535), (10, 247), (11, 575), (14, 747), (15, 653)]
Количество узлов: 41682


<TimeitResult : 2.92 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

## Тест №4

- Входной массив перемешан
- Операций поиска больше, чем операций добавления и удаления
- N = 50000

In [15]:
N = 50000
weights = [2, 5, 1]

In [16]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeAVL, N, weights=weights, shuffle=True)

Обход дерева: [(0, 356), (1, 486), (2, 833), (4, 179), (5, 987), (6, 555), (7, 667), (8, 266), (9, 758), (10, 691)]
Количество узлов: 33464


<TimeitResult : 4.21 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

In [17]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeRandom, N, weights=weights, shuffle=True)

Обход дерева: [(0, 506), (2, 892), (4, 974), (6, 922), (7, 41), (8, 557), (11, 296), (13, 948), (14, 981), (15, 394)]
Количество узлов: 33234


<TimeitResult : 9.47 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

In [18]:
%%timeit -r 1 -n 1 -o -q

tree = test(TreeSplay, N, weights=weights, shuffle=True)

Обход дерева: [(0, 135), (1, 497), (2, 77), (4, 890), (5, 344), (6, 673), (7, 393), (10, 988), (13, 536), (14, 189)]
Количество узлов: 33306


<TimeitResult : 5.09 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)>

## Выводы

|Номер теста|АВЛ|Random|Splay|
|:--------:|---------|--------------|--------------------|
|1|3.38|3.836|8.759|
|2|4.242|7.449|6.794|
|3|3.333|8.847|2.921|
|4|4.211|9.474|5.086|

- АВЛ-дерево показывает стабильный результат во всех тестах
- Рандомизированное дерево показывает себя лучше splay в случае отсортированного массива на входе, когда операций добавления больше, чем операций поиска и удаления. При таком варианте высока вероятность перебалансировки дерева перестановкой добавляемого узла в корень поддерева.
- Splay дерево начинает показывать себя лучше рандомизированного в случае отсортированного массива на входе, операций поиска больше, чем операций добавления и удаления. При таком варианте перебалансировка дерева происходит при помощи перестановки найденного узла в корень дерева, что позволяет также ускорить поиск часто запрашиваемых элементов.

In [19]:
# для сбора замеров времени в один массив
from IPython import get_ipython
from IPython.core.magics.execution import TimeitResult

RESULTS = []

def collect_results():
    for k, v in get_ipython().__dict__['user_ns']['Out'].items():
        if v.__class__ == TimeitResult:
            RESULTS.append(round(v.worst, 3))

In [20]:
collect_results()

In [21]:
for i in range(4):
    print(('|' + '|'.join(str(v) for v in RESULTS[i*3:3*(i+1)]) + '|'))

|3.38|3.836|8.759|
|4.242|7.449|6.794|
|3.333|8.847|2.921|
|4.211|9.474|5.086|
